drivers/platform/aducm3029/irq.c

```diff
@@ -0,0 +1,302 @@
+ * @param aducm_desc - Descriptor where the user callback is stared
+ * @param event - Event that generated the callback
+ * @param arg - Unused
+ */
+static void internal_callback(void *aducm_desc, uint32_t event, void *arg)
+{
+	struct aducm_irq_desc *desc = aducm_desc;
+
+	(void)arg;
+
+	switch (event) {
+	case ADI_XINT_EVENT_INT0:
+		if (desc->irq_handler[0])
+			desc->irq_handler[0]((void *)0);
+		break;
+	case ADI_XINT_EVENT_INT1:
+		if (desc->irq_handler[1])
+			desc->irq_handler[1]((void *)1);
+		break;
+	case ADI_XINT_EVENT_INT2:
+		if (desc->irq_handler[2])
+			desc->irq_handler[2]((void *)2);
+		break;
+	case ADI_XINT_EVENT_INT3:
+		if (desc->irq_handler[3])
+			desc->irq_handler[3]((void *)3);
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * @brief Initialized the controller for the ADuCM3029 external interrupts
+ *
+ * @param desc - Pointer where the configured instance is stored
+ * @param param - Configuration information for the instance
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_ctrl_init(struct irq_desc **desc,
+		      const struct irq_init_param *param)
+{
+	struct aducm_irq_desc		*aducm_desc;
+
+	if (!desc || !param || param->irq_id >= NB_INTERRUPT_CONTROLLERS ||
+	    initialized)
+		return FAILURE;
+
+	if (!(*desc = calloc(1, sizeof(**desc))))
+		return FAILURE;
+	if (!(aducm_desc = calloc(1, sizeof(*aducm_desc)))) {
+		free(*desc);
+		*desc = NULL;
+		return FAILURE;
+	}
+
+	(*desc)->extra = aducm_desc;
+	(*desc)->irq_id = param->irq_id;
+
+	adi_xint_Init(aducm_desc->irq_memory, ADI_XINT_MEMORY_SIZE);
+
+	initialized = 1;
+	return SUCCESS;
+}
+
+/**
+ * @brief Free the resources allocated by \ref irq_ctrl_init()
+ * @param desc - Interrupt controller descriptor.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_ctrl_remove(struct irq_desc *desc)
+{
+	if (!desc || !desc->extra || !initialized)
+		return FAILURE;
+
+	initialized = 0;
+	adi_xint_UnInit();
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Registers a generic IRQ handling function
+ * @param desc - Interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt
+ * @param irq_handler - Generic function to be registered. Will be called using
+ * as parameter the interrupt ID.
+ * @param dev_instance - Specify the trigger condition for the interrupt. To be
+ * one of the values from \ref irq_mode.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_register(struct irq_desc *desc, uint32_t irq_id,
+		     void (*irq_handler)(void *data), void *dev_instance)
+{
+	struct aducm_irq_desc *aducm_desc;
+	if (!desc || !desc->extra || !initialized || !irq_handler ||
+	    irq_id >= NB_EXT_INTERRUPTS)
+		return FAILURE;
+
+	aducm_desc = desc->extra;
+	aducm_desc->irq_handler[irq_id] = irq_handler;
+	aducm_desc->mode[irq_id] = (enum irq_mode)dev_instance;
+
+	adi_xint_RegisterCallback(id_map_event[irq_id], internal_callback,
```

Comment:
Should this be done in init only once rather than every time when registering a new callback?

Reply:
https://github.com/analogdevicesinc/no-OS/blob/201833ce661b164904d7a2de9b39dbe64946566f/include/irq.h#L94
It the implementation for the header function.

Reply:
I modify the comment of the function, maybe that was inaccurate.